File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/sensors/MagnetometerUncalibratedModule.java
// Copyright 2015-present 650 Industries. All rights reserved.

package versioned.host.exp.exponent.modules.api.sensors;

import android.hardware.Sensor;
import android.hardware.SensorEvent;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter;


public class MagnetometerUncalibratedModule extends SensorModuleBase {

  @Override
  public String getName() {
    return "ExponentMagnetometerUncalibrated";
  }

  public MagnetometerUncalibratedModule(ReactApplicationContext reactContext) {
    super(reactContext);
  }

  private static WritableMap eventToMap(SensorEvent sensorEvent) {
    WritableMap map = Arguments.createMap();
    map.putDouble("x", sensorEvent.values[0]);
    map.putDouble("y", sensorEvent.values[1]);
    map.putDouble("z", sensorEvent.values[2]);
    return map;
  }

  @Override
  public int getSensorType() {
    return Sensor.TYPE_MAGNETIC_FIELD_UNCALIBRATED;
  }

  @Override
  public void onSensorDataChanged(SensorEvent sensorEvent) {
    getReactApplicationContext().getJSModule(RCTDeviceEventEmitter.class).
        emit("magnetometerUncalibratedDidUpdate", eventToMap(sensorEvent));
  }

  @ReactMethod
  public void setUpdateInterval(int updateInterval) {
    super.setUpdateInterval(updateInterval);
  }

  @ReactMethod
  public void startObserving() {
    super.startObserving();
  }

  @ReactMethod
  public void stopObserving() {
    super.stopObserving();
  }
}

File: android/versioned-abis/expoview-abi48_0_0/src/main/java/abi48_0_0/expo/modules/sensors/modules/MagnetometerUncalibratedModule.kt
// Copyright 2015-present 650 Industries. All rights reserved.
package abi48_0_0.expo.modules.sensors.modules

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorManager
import android.os.Bundle
import abi48_0_0.expo.modules.interfaces.sensors.SensorServiceInterface
import abi48_0_0.expo.modules.interfaces.sensors.services.MagnetometerUncalibratedServiceInterface
import abi48_0_0.expo.modules.core.Promise
import abi48_0_0.expo.modules.core.interfaces.ExpoMethod

class MagnetometerUncalibratedModule(reactContext: Context?) : BaseSensorModule(reactContext) {
  override val eventName: String = "magnetometerUncalibratedDidUpdate"

  override fun getName(): String = "ExponentMagnetometerUncalibrated"

  override fun getSensorService(): SensorServiceInterface {
    return moduleRegistry.getModule(MagnetometerUncalibratedServiceInterface::class.java)
  }

  override fun eventToMap(sensorEvent: SensorEvent): Bundle {
    return Bundle().apply {
      putDouble("x", sensorEvent.values[0].toDouble())
      putDouble("y", sensorEvent.values[1].toDouble())
      putDouble("z", sensorEvent.values[2].toDouble())
    }
  }

  @ExpoMethod
  fun startObserving(promise: Promise) {
    super.startObserving()
    promise.resolve(null)
  }

  @ExpoMethod
  fun stopObserving(promise: Promise) {
    super.stopObserving()
    promise.resolve(null)
  }

  @ExpoMethod
  fun setUpdateInterval(updateInterval: Int, promise: Promise) {
    super.setUpdateInterval(updateInterval)
    promise.resolve(null)
  }

  @ExpoMethod
  fun isAvailableAsync(promise: Promise) {
    val mSensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    val isAvailable = mSensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD_UNCALIBRATED) != null
    promise.resolve(isAvailable)
  }
}
